File: advent-of-code/2020/1/solver.js
import { readFileSync } from 'fs';
import { Seq } from 'immutable';
import { EOL } from 'os';

const add = (x, y) => x + y
const sum = (numbers) => numbers.reduce(add)

const data = readFileSync('expenses.txt', 'utf8')
    .split(EOL)
    .map(x => Number.parseInt(x))
    .sort((a, b) => a - b);

const ascending = Seq(data);
const descending = Seq(ascending).reverse();
console.log(
    descending.flatMap(x =>
        ascending.map(y => [x, y])
            .takeUntil(xy => sum(xy) > 2020))
        .flatMap(xy =>
            ascending.map(z => [...xy, z])
                .takeUntil(xy => sum(xy) > 2020))
        .find(xyz => sum(xyz) === 2020)
)

File: advent-of-code/2020/1/haskell/solver.hs
{-
resolver: lts-18.9
-}
import           Data.List

main = do
    content <- readFile "../input/expenses.txt"
    let nums = map readInt $ lines content
    print $ solve 2020 $ sort nums
    return ()

readInt :: String -> Int
readInt = read

solve target numbers = solve' target $ sort numbers
solve' target numbers = head
    [ x : yz
    | yz <-
        [ [y, z]
        | z <- reverse numbers
        , y <- takeWhile (\n -> n + z < target) numbers
        ]
    , x <- takeWhile (\n -> sum (n : yz) <= target) numbers
    , sum (x : yz) == target
    ]
